Core/src/com/infiniteautomation/mango/util/script/CompiledMangoJavaScript.java

```diff
@@ -0,0 +1,320 @@
+
+/**
+ * Container for Compiled Java Scripts to aid in state management 
+ * of the various scopes and to help simplify user code.
+ * 
+ * Usage:
+ * 1. Constructe compiled script
+ * 2. call compile()
+ * 3. call initialize() (anytime the context changes but now while executing the script)
+ * 4. call execute() when you desire to compute a result from the script
+ * 
+ * @author Terry Packer
+ *
+ */
+public class CompiledMangoJavaScript {
+    
+    private CompiledScript compiledScript;
+    
+    private final ScriptPointValueSetter setter;
+    private final ScriptLog log;
+    
+    private final Map<String, Object> additionalContext;
+    private final List<ScriptUtility> additionalUtilities;
+    private final List<JsonImportExclusion> importExclusions;
+    private final boolean testRun;
+    private final MangoJavaScriptService service;
+
+    private final MangoJavaScriptResult result;
+    private final List<ScriptUtility> utilities;
+    private final PermissionHolder user;
+    
+    //State settings
+    private boolean initialized;
+    
+    /**
+     * Create a compiled script container for testing
+     * 
+     * NOTE: Script engine and utilities are run using permissions of vo
+     * 
+     * @param vo
+     * @param log
+     * @param result
+     * @param service
+     */
+    public CompiledMangoJavaScript(MangoJavaScript vo, ScriptLog log, MangoJavaScriptResult result, MangoJavaScriptService service) {
+        this.log = log;
+        this.additionalContext = vo.getAdditionalContext() == null ? Collections.emptyMap() : vo.getAdditionalContext();
+        this.additionalUtilities = vo.getAdditionalUtilities() == null ? Collections.emptyList() : vo.getAdditionalUtilities();
+        this.importExclusions = Collections.emptyList();
+        this.testRun = true;
+        this.service = service;
+        this.result = result;
+        this.utilities = new ArrayList<>();
+        this.user = vo.getPermissions();
+        this.setter = service.createValidationSetter(result, user);
+    }
+    
+    /**
+     * Create a compiled script container to run live
+     * 
+     * NOTE: Script engine and utilities are run using permissions of vo
+     * 
+     * @param vo
+     * @param setter
+     * @param log
+     * @param result
+     * @param service
+     */
+    public CompiledMangoJavaScript(MangoJavaScript vo, ScriptPointValueSetter setter, ScriptLog log, MangoJavaScriptResult result, MangoJavaScriptService service) {
+        this.log = log;
+        this.additionalContext = vo.getAdditionalContext() == null ? Collections.emptyMap() : vo.getAdditionalContext();
+        this.additionalUtilities = vo.getAdditionalUtilities() == null ? Collections.emptyList() : vo.getAdditionalUtilities();
+        this.importExclusions = Collections.emptyList();
+        this.testRun = true;
+        this.service = service;
+        this.result = result;
+        this.utilities = Collections.emptyList();
```

Comment:
I'm not sure Collections.emptyList() is what's desired here. Seems to me the initialize methods in the service will promptly attempt to mutate it, but have only been reading, didn't try to evoke

Reply:
Agreed.  Done.